### PR TITLE
`<future>`: Support `future<const int>` and `promise<const int>`

### LINKED ISSUE #2599: `<future>`: Consider supporting `promise<const int>` and `future<const int>`
Followup to #2488 fixed by #2568. Nothing in the Standard appears to forbid `promise<const int>` and `future<const int>`, so they should work. It should be possible to slightly refine the new code added by #2568 to handle `const` qualifiers, but additional changes may be necessary, and test coverage is definitely necessary.

Noticed by @CaseyCarter in https://github.com/microsoft/STL/pull/2568#discussion_r813608299 .

### PR DESCRIPTION
**Pull request by frederick-vs-ja**
Default-initializable but not assignable types are specially handled.

Fixes #2599. Also addresses #1190. Updating the referenced document number to [N4917](https://wg21.link/n4917) with section numbers dropped (towards #182).

Unblocking one libcxx test:
- `std/thread/futures/futures.promise/set_rvalue.pass.cpp`

### CODE REVIEW

**Review by strega-nil-ms — CHANGES_REQUESTED** [review]
I don't know if this change is possible, unfortunately, for non-trivially destructible types.

File: stl/inc/future [hunk 1]
```diff
@@ -211,7 +217,7 @@ public:
             _Cond._Unregister(_Mtx);
         }
 
-        if constexpr (!is_default_constructible_v<_Ty>) {
+        if constexpr (!_Is_default_newable<_Ty>) {
```
[the remark this answers is not in the document]
Comment (frederick-vs-ja) [reply stl/inc/future:220]: I think this PR doesn't change any behavior that currently compiles. If I get it right, `future<const vector<int>>` won't have different behavior in this example, because `const vector<int>` satisfies `_Is_default_newable`.

File: stl/inc/future [hunk 2]
```diff
@@ -308,11 +314,18 @@ public:
...
-            _Result = _STD forward<_Ty2>(_Val);
+        if constexpr (_Is_default_newable<_Ty>) {
+            // TRANSITION, incorrectly assigns _Result when _Ty is default initializable and assignable
+            if constexpr (is_assignable_v<_Ty&, _Ty2>) {
+                _Result = _STD forward<_Ty2>(_Val);
+            } else {
+                if constexpr (!is_trivially_destructible_v<_Ty>) {
+                    _Result.~_Ty();
+                }
+                ::new (_STD _Voidify_iter(_STD addressof(_Result))) _Ty(_STD forward<_Ty2>(_Val));
```
[the remark this answers is not in the document]
Comment (frederick-vs-ja) [reply stl/inc/future:325]: This one is actually problematic. When `_Is_default_newable<_Ty>` is `true`, we must ensure that the `_Ty` object is living before destroying `_Associated_state`. I'll try to fix this case.

File: stl/inc/future [hunk 3]
```diff
@@ -299,20 +305,62 @@ public:
...
             return _Result._Held_value;
         }
     }
 
+    // TRANSITION, keeps _Result living
+    struct _Reconstruction_guard {
+        _Ty* _Res_addr;
+
+        _Reconstruction_guard& operator=(const _Reconstruction_guard&) = delete;
+        _Reconstruction_guard& operator=(_Reconstruction_guard&&) = delete;
```
Comment (fsb4000) [inline stl/inc/future:320]: clang-format wants that these two `= delete` would be one under another.

File: stl/inc/future (hunk 2 above)
[the remark this answers is not in the document]
Comment (frederick-vs-ja) [reply stl/inc/future]: I guess I've made a suitable fix which keeps that libcxx test passed. In some edge cases `set_value` etc. would call `terminate` due to the exception thrown by a default ctor, but these cases don't even compile currently.

**Comment by StephanTLavavej** [comment]
Thanks for looking into this issue. We talked about it at the weekly maintainer meeting, and we decided that because we can't support all `const` types here, we shouldn't attempt to support *some* `const` types. (It would be very difficult to explain to users what works and what doesn't - I'm personally having trouble wrapping my brain around it after several minutes of looking at this after returning from vacation :joy_cat: :brain:.)

We've updated the [vNext wiki](https://github.com/microsoft/STL/wiki/vNext-Planning) to record that this should be supported when we can overhaul the representation.

**Comment by frederick-vs-ja** [comment]
I think I was trying to cover all const types, except that default-initializable types may result in non-standard behavior due to ABI compatibility (which is pre-existing).
